src/plugins/timeseries-plugin.ts

```diff
@@ -0,0 +1,215 @@
+    height: number;
+    padding: number;
+    color: string;
+    lineStyle: 'step' | 'interpolate';
+};
+
+export const defaultCPUPluginStyles: TimeseriesPluginStylesNoOptional = {
+    height: 68,
+    padding: 5,
+    lineStyle: 'interpolate',
+    color: 'red',
+};
+
+export class TimeseriesPlugin extends UIPlugin<TimeseriesPluginStylesNoOptional> {
+    height: number;
+    data: TimeseriesPoint[];
+    summary: TimeseriesPointsSummary | null = null;
+    override styles: TimeseriesPluginStylesNoOptional;
+
+    constructor({ name, data, styles }: { name: string; data: TimeseriesPoint[]; styles?: TimeseriesPluginStyles }) {
+        super(name);
+
+        this.styles = { ...defaultCPUPluginStyles, ...(styles ?? {}) };
+        this.data = [];
+        this.name = name;
+
+        this.height = this.styles.height;
+
+        this.setData(data);
+    }
+
+    override init(renderEngine: OffscreenRenderEngine, interactionsEngine: SeparatedInteractionsEngine) {
+        super.init(renderEngine, interactionsEngine);
+
+        this.interactionsEngine.on('change-position', this.handlePositionChange.bind(this));
+        this.interactionsEngine.on('up', this.handleMouseUp.bind(this));
+    }
+
+    handlePositionChange(position: { deltaX: number }) {
+        const startPositionX = this.renderEngine.parent.positionX;
+
+        this.interactionsEngine.setCursor('grabbing');
+
+        this.renderEngine.tryToChangePosition(position.deltaX);
+
+        if (startPositionX !== this.renderEngine.parent.positionX) {
+            this.renderEngine.parent.render();
+        }
+    }
+
+    handleMouseUp() {
+        this.interactionsEngine.clearCursor();
+    }
+
+    setData(data: TimeseriesPoint[]) {
+        this.data = data;
+
+        let min = Number.MAX_VALUE;
+        let max = Number.MIN_VALUE;
+        let first = Number.MAX_VALUE;
+        let last = Number.MIN_VALUE;
+
+        this.data.forEach(([timestamp, value]) => {
+            if (value < min) {
+                min = value;
+            }
+            if (value > max) {
+                max = value;
+            }
+
+            if (timestamp < first) {
+                first = timestamp;
+            }
+            if (timestamp > last) {
+                last = timestamp;
+            }
+        });
+
+        this.summary = {
+            min,
+            max,
+            first,
+            last,
+        };
+    }
+
+    calcRect(start: number, duration: number, isEnd: boolean) {
+        const w = duration * this.renderEngine.zoom;
+
+        return {
+            x: this.renderEngine.timeToPosition(start),
+            w: isEnd ? (w <= 0.1 ? 0.1 : w >= 3 ? w - 1 : w - w / 3) : w,
+        };
+    }
+
+    findClosestDataPoint(timestamp: number): TimeseriesPoint {
+        return [timestamp, -1];
+    }
+
+    override renderTooltip() {
+        const mouse = this.interactionsEngine.getGlobalMouse();
+        //QUESTION: how do i convert mouse.x to position
```

Comment:
@pyatyispyatil  - How to i convert mouse.x position to timestamp?

@pyatyispyatil - the tooltip kind of works - but it is alittle flaky - sometimes it appears, sometimes it doesn't. there seems to be something not quite right

Reply:
You can convert pixels to time like this
```ts
this.renderEngine.pixelToTime(this.interactionsEngine.mouse.x) + this.renderEngine.min
```

Reply:
About glitches, I understand. I made the use of the hover a bit incomprehensible. I will improve this point further, but for now, you need to subscribe to hover and save the region the same way as in other plugins and use it for validation in the tooltip's render method. If there is no region, do not render the tooltip and return false.
<img width="825" alt="image" src="https://user-images.githubusercontent.com/4976306/213403100-4323a21c-6f58-4aae-9ddf-6112b5276915.png">
<img width="644" alt="image" src="https://user-images.githubusercontent.com/4976306/213403168-867019f8-7100-4338-ba72-ee13b47f515d.png">

Reply:
Yes, I found another hover related bug. If you move the cursor from one of the flame chart blocks directly to your chart, the flame chart tooltip remains. This is an engine bug - I'll deal with it today, you don't have to worry about it.

Reply:
Fixed in https://github.com/pyatyispyatil/flame-chart-js/commit/b2f5e9d335a89d0e78faa3c425b729bb60d68995

Reply:
great - i found that bug too!

Reply:
> I think have done everything you asked now;

> Please note - if you tooltip over the flame-chart first the tooltip is broken... but as long as you don't it works just fine...

> I think this is a bug somewhere else in the render engine? or a lack of undertanding on my part.

This bug is on your side. I mentioned it above https://github.com/pyatyispyatil/flame-chart-js/pull/44#discussion_r1080995355
Your tooltip remains because you don't check the hovered region.

https://github.com/pyatyispyatil/flame-chart-js/pull/44#discussion_r1081142303 - this message about another one.)